control-plane/api-gateway/controllers/gatewayclass_controller.go

```diff
@@ -0,0 +1,214 @@
+	gatewayClassFieldIndex       = "__gatewayclass"
+)
+
+// GatewayClassReconciler reconciles a GatewayClass object.
+// The GatewayClass is responsible for defining the behavior of API gateways
+// which reference the given class.
+type GatewayClassReconciler struct {
+	ControllerName string
+	Log            logr.Logger
+	client.Client
+}
+
+// Reconcile handles the reconciliation loop for GatewayClass objects.
+func (r *GatewayClassReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := r.Log.WithValues("gatewayClass", req.NamespacedName)
+
+	gc := &gwv1beta1.GatewayClass{}
+
+	err := r.Client.Get(ctx, req.NamespacedName, gc)
+	if err != nil {
+		log.Error(err, "unable to get GatewayClass")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if string(gc.Spec.ControllerName) != r.ControllerName {
+		// This GatewayClass is not for this controller.
+		_, err := RemoveFinalizer(ctx, r.Client, gc, GatewayClassFinalizer)
+		if err != nil {
+			log.Error(err, "unable to remove finalizer")
+			return ctrl.Result{}, err
+		}
+
+		return ctrl.Result{}, nil
+	}
+
+	if !gc.ObjectMeta.DeletionTimestamp.IsZero() {
+		// We have a deletion request. Ensure we are not in use.
+		used, err := r.isGatewayClassInUse(ctx, gc)
+		if err != nil {
+			log.Error(err, "unable to check if GatewayClass is in use")
+			return ctrl.Result{}, err
+		}
+		if used {
+			log.Info("GatewayClass is in use, cannot delete")
+			return ctrl.Result{}, nil
+		}
+		// Remove our finalizer.
+		if _, err := RemoveFinalizer(ctx, r.Client, gc, GatewayClassFinalizer); err != nil {
+			log.Error(err, "unable to remove finalizer")
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	// We are creating or updating the GatewayClass.
+	didUpdate, err := EnsureFinalizer(ctx, r.Client, gc, GatewayClassFinalizer)
+	if err != nil {
+		log.Error(err, "unable to add finalizer")
+		return ctrl.Result{}, err
+	}
+	if didUpdate {
+		// We updated the GatewayClass, requeue to avoid another update.
+		return ctrl.Result{}, nil
+	}
+
+	// Check that the parametersRef is valid.
+	pr := gc.Spec.ParametersRef
+	if pr != nil {
+		if pr.Kind != v1alpha1.GatewayClassConfigKind {
+			_, err := r.ensureStatus(ctx, gc, InvalidParameters, fmt.Sprintf("Incorrect type for parametersRef. Expected GatewayClassConfig, got %q.", pr.Kind))
+			if err != nil {
+				log.Error(err, "unable to update status")
+			}
+			return ctrl.Result{}, err
+		}
+
+		err := r.Client.Get(ctx, types.NamespacedName{Name: pr.Name}, &v1alpha1.GatewayClassConfig{})
+		if k8serrors.IsNotFound(err) {
+			_, err := r.ensureStatus(ctx, gc, InvalidParameters, fmt.Sprintf("GatewayClassConfig not found %q.", pr.Name))
+			if err != nil {
+				log.Error(err, "unable to update status")
+			}
+			return ctrl.Result{}, err
+		}
+		if err != nil {
+			log.Error(err, "unable to fetch GatewayClassConfig")
+			return ctrl.Result{}, err
+		}
+	}
+
+	return ctrl.Result{}, nil
```

Comment:
One more change we'll need here, if we have no errors in validation, we should be setting the `GatewayClass` status to have an `Accepted == True` condition.